### DISCUSSION TITLE
How to use xmeshfem2D with external model ?

### DISCUSSION THREAD

**Discussion by ICNSK** [SPECFEM/specfem3d#1692] [SPECFEM2D General]
Hi,
I am brand new user and trying to learn how to run xspecfem2D my research work. In the past I used finite difference simulators, which were limited to acoustic  mode.   I have input model for finite difference simulations, which  is regularly samples binary file N1 x N2 and not sure how to convert it to mesh and model which could be used in xspecfem2D. 

**I was planning to use xmeshfem2D but could not find an example of how to use it with external mode.**

Any information will be greatly appreciated, Thank you in advance

**Comment by mnagaso** [comment]
Hi,

For using the exernal model, you can modify this code and run with `Par_file/MODEL = external` :
https://github.com/SPECFEM/specfem2d/blob/master/src/specfem2D/define_external_model.f90

In your case, you can read your N1 x N2 model file in `define_external_model` subroutine, then you need to interpolate  the grid coordinate into the specfem's nodes' position.

I hope this would help you,

**Reply by ICNSK** [reply]
Dear @mnagaso 

Thank you very much for pointing me in the good direction ! I have limited experience programming in Fortran, but hopefully will be able to modify the `define_external_model `subroutine to implement functionality i need.  As of now i am using 2 stage approach (generate `proc***_rho_vp_vs.dat` with "dummy" run with  `Par_file/SAVE_MODEL = ascii `, unfortunately file is  written at the end of simulation,  than modify it using my binary file with material properties manually and run code second time with updated model using `Par_file/Model = ascii `)

**Reply by mnagaso** [reply]
Hi @ICNSK 

Yes, your approach is also an alternative choice! I sometimes do that way actually if the mesh size is not so large. For the large mesh, it is better to do this with define_external_model. But the results should be identical with whichever way we use👍
